### PR TITLE
Add support of cross-building  the engine for ARM64 Linux Host

### PR DESCRIPTION
**Pull request by HidenoriMatsubayashi**
## Description

Added cross-building Flutter Engine for ARM64 Linux platforms. This PR is part of ARM64 Linux support in the Flutter SDK.

## Related PRs
 - [ ] https://github.com/flutter/flutter/pull/61221
 - [x] https://github.com/flutter/engine/pull/20254

## Related Issues
* https://github.com/flutter/flutter/issues/56992
* https://github.com/flutter/flutter/issues/60678

### 1. Build Test
First, I confirmed that Flutter Engine can be built for x64 and arm64. The procedure is as follows. All are operations on x64 Linux Host.

Get source code and setup build enviroment.

```
create .gclient file.
$ gclient sync
$ cd src
````

Cross-build for ARM64 Linux

```
$ ./flutter/tools/gn --target-os linux --linux-cpu arm64 --runtime-mode release
$ ninja -C out/linux_release_arm64
-> Passed!

$ ./flutter/tools/gn --target-os linux --linux-cpu arm64 --runtime-mode debug
$ ninja -C out/linux_debug_arm64
-> Passed!
```

Build for x64 Linux and Regression test.

```
$ ./flutter/tools/gn --unoptimized
$ ninja -C out/host_debug_unopt
-> Passed!

$ ./flutter/testing/run_tests.sh
-> All Tests Passed!
```

### 2. Run Flutter SDK and app on ARM64 Linux (Jetson Nano)

Second, using the artifacts for ARM64 built by the above procedure, I have confirmed that the Flutter SDK and Flutter app can work fine on Jetson Nano (ARM64 CPU).

All are operations on ARM64 Linux Host (Jetson Nano).

```
$ git clone https://github.com/flutter/flutter.git
$ export PATH=$PATH:{your working dir path}/flutter/bin
$ flutter doctor
   ... Error happened.
```

Copy `dart-sdk` which was locally built to `${your installpath}/bin/cache/`.

```
$ flutter doctor
$ flutter config --enable-linux-desktop
$ flutter create sample
$ cd sample
$ flutter build linux
   ... Error happened.
```

Copy the Flutter SDK artifacts (build data) to `${your installpath}/bin/cache/artifacts/engine` directory manually.

 -  `out/linux_debug_arm64/flutter_patched_sdk` to `${your installpath}/bin/cache/artifacts/engine/common/flutter_patched_sdk`
 -  `out/linux_debug_arm64/flutter_linux, gen/const_finder.dart.snapshot, flutter_tester, font-subset, gen/frontend_server.dart.snapshot, icudtl.dat, libflutter_linux_gtk.so, gen/flutter/lib/snapshot/isolate_snapshot.bin, gen/flutter/lib/snapshot/vm_isolate_snapshot.bin` to `${your installpath}/bin/cache/artifacts/engine/artifacts/engine/linux-x64`

-  `out/linux_release_arm64/flutter_patched_sdk` to `${your installpath}/bin/cache/artifacts/engine/common/flutter_patched_sdk_product`
 -  `out/linux_release_arm64/flutter_linux, libflutter_linux_gtk.so` to `${your installpath}/bin/cache/artifacts/engine/artifacts/engine/linux-x64-release`
 -  `out/linux_release_arm64/clang_x64/gen_snapshot` to `${your installpath}/bin/cache/artifacts/engine/artifacts/engine/linux-x64-release`

* For simple operation check, `${your installpath}/bin/cache/artifacts/engine/artifacts/engine/linux-x64-profile` is omitted.

```
( $ flutter build linux --verbose )
$ flutter run -d linux
```

Screenshot from 2020-08-05 12-52-42

### CODE REVIEW

File: build/config/sysroot.gni [hunk 1]
```diff
@@ -27,8 +27,12 @@ if (current_toolchain == default_toolchain && target_sysroot != "") {
 } else if (is_linux && !is_chromeos) {
   if (current_cpu == "mipsel") {
     sysroot = rebase_path("//mipsel-sysroot/sysroot")
-  } else if (use_default_linux_sysroot && !is_fuchsia && current_cpu == "x64") {
-    sysroot = rebase_path("//build/linux/debian_sid_amd64-sysroot")
+  } else if (use_default_linux_sysroot && !is_fuchsia) {
+    if (current_cpu == "x64") {
+      sysroot = rebase_path("//build/linux/debian_sid_amd64-sysroot")
+    } else {
+      sysroot = rebase_path("//build/linux/debian_sid_arm64-sysroot")
```
Comment (stuartmorgan) [inline build/config/sysroot.gni:34]: I'm a little confused by the scope of this PR; the title says "Host", but this change appears to be specific to building the target artifacts. And the "Tests" section (which isn't actually discussing tests at all) appears to be showing a manual end-to-end process of some kind but I'm not clear on whether it's host, target, or both.

Please update the PR description to clearly describe what exactly this PR is for. It appears to me from the changes that the answer is that it's for cross-compiling the engine on Linux.

Comment (HidenoriMatsubayashi) [reply build/config/sysroot.gni:34]: Thank you.
It's my mistake. target_cpu is correct.. Please wait for a while.

Comment (stuartmorgan) [reply build/config/sysroot.gni:34]: `current_cpu` looks correct to me in this usage; my confusion was around the description of the PR vs what it seems to be doing.

Comment (HidenoriMatsubayashi) [reply build/config/sysroot.gni:34]: I updated the description. Could you check that?

File: build/toolchain/linux/BUILD.gn [hunk 2]
```diff
@@ -100,7 +100,7 @@ gcc_toolchain("clang_arm64") {
   nm = "nm"
   ar = "ar"
   ld = cxx
-  strip = "strip"
+  strip = "${compiler_prefix}$prefix/llvm-strip"
```
Comment (stuartmorgan) [inline build/toolchain/linux/BUILD.gn:103]: Why is this needed? Also, what are the implications for x64 for changing from `string` to `llvm-strip`?

Comment (HidenoriMatsubayashi) [reply build/toolchain/linux/BUILD.gn:103]: Thank you for your response.

> Why is this needed?

This is to fix the following error. We have to use the binaries under `buildtools/linux-x64/clang/bin/`. I think that originally wrong. [dart-sdk](https://github.com/dart-lang/sdk/blob/master/build/toolchain/linux/BUILD.gn) is helpful. I guess it's better to modify `readelf`, `nm`, and `ar` as well...

```
[4439/6380] LINK ./font-subset
FAILED: font-subset exe.unstripped/font-subset
../../buildtools/linux-x64/clang/bin/clang++ -Wl,--fatal-warnings -fPIC -Wl,-z,noexecstack -Wl,-z,now -Wl,-z,relro -Wl,-z,defs -pthread --target=aarch64-linux-gnu --sysroot=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot -L/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/local/lib/aarch64-linux-gnu\ -Wl,-rpath-link=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/local/lib/aarch64-linux-gnu\ -L/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/lib/aarch64-linux-gnu\ -Wl,-rpath-link=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/lib/aarch64-linux-gnu\ -L/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/lib/aarch64-linux-gnu\ -Wl,-rpath-link=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/lib/aarch64-linux-gnu -Wl,-O2 -Wl,--gc-sections -Wl,--as-needed -Wl,-rpath=\$ORIGIN/ -Wl,-rpath-link= -Wl,--disable-new-dtags  -o ./exe.unstripped/font-subset -Wl,--build-id=sha1 -Wl,--start-group @./font-subset.rsp  -Wl,--end-group  -ldl  && strip --strip-unneeded -o ./font-subset ./exe.unstripped/font-subset
strip: Unable to recognise the format of the input file `./exe.unstripped/font-subset'
[4444/6380] LINK ./dart_precompiled_runtime_product
FAILED: dart_precompiled_runtime_product exe.unstripped/dart_precompiled_runtime_product
../../buildtools/linux-x64/clang/bin/clang++ -rdynamic -Wl,--fatal-warnings -fPIC -Wl,-z,noexecstack -Wl,-z,now -Wl,-z,relro -Wl,-z,defs -pthread --target=aarch64-linux-gnu --sysroot=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot -L/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/local/lib/aarch64-linux-gnu\ -Wl,-rpath-link=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/local/lib/aarch64-linux-gnu\ -L/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/lib/aarch64-linux-gnu\ -Wl,-rpath-link=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/lib/aarch64-linux-gnu\ -L/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/lib/aarch64-linux-gnu\ -Wl,-rpath-link=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/lib/aarch64-linux-gnu -Wl,-O2 -Wl,--gc-sections -Wl,--as-needed -Wl,-rpath=\$ORIGIN/ -Wl,-rpath-link= -Wl,--disable-new-dtags  -o ./exe.unstripped/dart_precompiled_runtime_product -Wl,--build-id=sha1 -Wl,--start-group @./dart_precompiled_runtime_product.rsp  -Wl,--end-group  -ldl -lpthread  && strip --strip-unneeded -o ./dart_precompiled_runtime_product ./exe.unstripped/dart_precompiled_runtime_product
strip: Unable to recognise the format of the input file `./exe.unstripped/dart_precompiled_runtime_product'
[4446/6380] LINK ./gen_snapshot_product
FAILED: gen_snapshot_product exe.unstripped/gen_snapshot_product
../../buildtools/linux-x64/clang/bin/clang++ -rdynamic -Wl,--fatal-warnings -fPIC -Wl,-z,noexecstack -Wl,-z,now -Wl,-z,relro -Wl,-z,defs -pthread --target=aarch64-linux-gnu --sysroot=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot -L/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/local/lib/aarch64-linux-gnu\ -Wl,-rpath-link=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/local/lib/aarch64-linux-gnu\ -L/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/lib/aarch64-linux-gnu\ -Wl,-rpath-link=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/lib/aarch64-linux-gnu\ -L/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/lib/aarch64-linux-gnu\ -Wl,-rpath-link=/root/flutter_work/build_engine/src/build/linux/debian_sid_arm64-sysroot/usr/lib/aarch64-linux-gnu -Wl,-O2 -Wl,--gc-sections -Wl,--as-needed -Wl,-rpath=\$ORIGIN/ -Wl,-rpath-link= -Wl,--disable-new-dtags  -o ./exe.unstripped/gen_snapshot_product -Wl,--build-id=sha1 -Wl,--start-group @./gen_snapshot_product.rsp  -Wl,--end-group  -ldl -lpthread  && strip --strip-unneeded -o ./gen_snapshot_product ./exe.unstripped/gen_snapshot_product
strip: Unable to recognise the format of the input file `./exe.unstripped/gen_snapshot_product'
[4452/6380] CXX clang_x64/obj/third_party/dart/runtime/lib/libdart_lib_jit.ffi_dynamic_library.o
ninja: build stopped: subcommand failed.
```

> Also, what are the implications for x64 for changing from string to llvm-strip?

No implications for x64 because `clang_x64` is used for self-building for x64, and the commands installed on a Host are used.

```
cc_toolchain("clang_x64") {
  if (use_clang_type_profiler) {
    prefix = rebase_path("//third_party/llvm-allocated-type/Linux_x64/bin",
                         root_build_dir)
  } else {
    prefix = rebase_path("//buildtools/linux-x64/clang/bin", root_build_dir)
  }
  cc = "${compiler_prefix}$prefix/clang"
  cxx = "${compiler_prefix}$prefix/clang++"

  readelf = "readelf"
  nm = "nm"
  ar = "ar"
  ld = cxx
  strip = "strip"

  toolchain_cpu = "x64"
  toolchain_os = "linux"
  is_clang = true
}
```

Comment (stuartmorgan) [reply build/toolchain/linux/BUILD.gn:103]: > No implications for x64 because clang_x64 is used for self-building for x64

Sorry, I didn't expand the context to see that it was in an arm-specific path.

Comment (stuartmorgan) [reply build/toolchain/linux/BUILD.gn:103]: > dart-sdk is helpful

Interestingly they don't seem to use llvm-strip though; they just do prefixing in that file. Do you know why it's different?

Comment (HidenoriMatsubayashi) [reply build/toolchain/linux/BUILD.gn:103]: Thank you so much. Please give me some time to think and investigate about that.

Comment (HidenoriMatsubayashi) [reply build/toolchain/linux/BUILD.gn]: I modified it to use llvm_objcopy for now.

**Comment by HidenoriMatsubayashi** [comment]
When I change from `nm` to `llvm-nm` command, the error happened.

```
  #nm = "nm"
  nm = "${prefix}/llvm-nm"
```

### The error message

```
$ ninja -C out/linux_release_arm64
ninja: Entering directory `out/linux_release_arm64'
[4664/6410] SOLINK ./libflutter_engine.so
llvm-nm: for the --format option: Cannot find option named 'p'!
[4683/6410] SOLINK ./libflutter_linux_gtk.so
llvm-nm: for the --format option: Cannot find option named 'p'!
[6410/6410] STAMP obj/default.stamp
```

**Comment by HidenoriMatsubayashi** [comment]
`-f p` is incorrect. This is the bug. `-f posix` is correct.

### [build/toolchain/gcc_toolchain.gni](https://github.com/flutter/buildroot/blob/236a74613244aa98e45ea0e2bffd16d03adb9286/build/toolchain/gcc_toolchain.gni#L172)
```
toc_command = "{ $readelf -d $sofile | grep SONAME ; $nm -gD -f p $sofile | cut -f1-2 -d' '; } > $temporary_tocname"
```

### $ man nm

```
      -f format
       --format=format
           Use the output format format, which can be "bsd", "sysv", or "posix".  The default is "bsd".  Only the first
           character of format is significant; it can be either upper or lower case.
```

File: build/toolchain/linux/BUILD.gn (hunk 2 above)
Comment (HidenoriMatsubayashi) [reply build/toolchain/linux/BUILD.gn]: @stuartmorgan 

I fixed. Could you review the source code in your free time?

**Review by stuartmorgan — APPROVED** [review]
Sorry, I forgot this part of the engine work was still outstanding. LGTM.

It looks like Dart is using llvm-objcopy for clang_x64 as well, so we should investigate whether to unify that at some point on our side, but limiting the changes to just arm64 here is the safer option.

**Comment by stuartmorgan** [comment]
Do you have a test PR in flutter/engine that uses this commit for the buildroot DEPS that is green, so we can be sure it's safe to land?

**Comment by HidenoriMatsubayashi** [comment]
> It looks like Dart is using llvm-objcopy for clang_x64 as well, so we should investigate whether to unify that at some point on our side

I agree with you. I think it is better to use LLVM commands as well Dart-sdk if you have no specific reasons. The Linux desktop version is still the alpha version, and if you want to change it, I think it's better before the beta version. So, I'll investigate about clang_x64 after this work.

> Do you have a test PR in flutter/engine that uses this commit for the buildroot DEPS that is green

I created the PR (https://github.com/flutter/engine/pull/22284) just now.

**Comment by stuartmorgan** [comment]
> I created the PR ([flutter/engine#22284](https://github.com/flutter/engine/pull/22284)) just now.

Thanks! Everything is green there, so I'll go ahead and land this.

**Comment by HidenoriMatsubayashi** [comment]
Thanks.
